### ISSUE TITLE
ssue 429 error code after 12h when calling "login.microsoftonline.../well-known" to call Azure SQL Database from Azure App Service

### ISSUE THREAD

**Issue by mhi59** [Azure/azure-sdk-for-net#47511]
Hello

I have an Azure App Service that is called multiple times per second. This App Service connects to an Azure SQL Database using a managed identity. The application works correctly for about 12 hours, but then I start seeing calls to “login.microsoftonline…/well-known…” that either time out or take a very long time.

This causes significant slowdowns because subsequent incoming requests also seem to try calling this URL to retrieve a refresh token, and all of them then take more than 30 seconds to complete and i have 429 http error code.

I don’t have any specific configuration in the code since, as far as I know, everything related to the refresh token should be handled automatically by the library.

Does someone had this problem too? Thank you

```c#
private static void ConfigureAzureIdentityAuthentication(this IServiceCollection services, IConfiguration configuration)
{
    var userAssignedManagedIdentity = configuration. GetValue<string?>("UserAssignedManagedIdentity");

//Managed Identities handlers - User Assigned Managed Identity
    services. AddAzureClients(clientBuilder =>
    {
        if (!string. IsNullOrEmpty(userAssignedManagedIdentity))
        {
            DefaultAzureCredential defaultAzureCredential = new(
                new DefaultAzureCredentialOptions
                {
                    ManagedIdentityClientId = userAssignedManagedIdentity
                });

clientBuilder.UseCredential(defaultAzureCredential);
        }
    });
}
```

 
![Image](https://github.com/user-attachments/assets/b135ec45-5418-4e23-bb4a-2632399d37b7)
![Image](https://github.com/user-attachments/assets/498206bd-6d3b-4678-8613-624922d1efad)
![Image](https://github.com/user-attachments/assets/f3bf98bf-8833-496a-94d6-6784c2ac3577)

**Comment by mhi59** [comment]
Thank you very much for your quick response @christothes 
I am using EF Core, and I configure it as follows:

```
services.AddDbContext<AppDbContext>(options => options.UseSqlServer(azureSqlOptions.ConnectionString));
```

My application communicates with BlobStorage, AzureBus, and AzureSqlDatabase using Managed Identities.

I have configured the Managed Identity like this:

```
private static void ConfigureAzureIdentityAuthentication(this IServiceCollection services, IConfiguration configuration)
{
    var userAssignedManagedIdentity = configuration.GetValue<string?>("UserAssignedManagedIdentity");

    // Managed Identities handlers - User Assigned Managed Identity
    services.AddAzureClients(clientBuilder =>
    {
        if (!string.IsNullOrEmpty(userAssignedManagedIdentity))
        {
            DefaultAzureCredential defaultAzureCredential = new(
                new DefaultAzureCredentialOptions
                {
                    ManagedIdentityClientId = userAssignedManagedIdentity,
                    ExcludeSharedTokenCacheCredential = false
                });

            clientBuilder.UseCredential(defaultAzureCredential);
        }
    });
}

```

Isn't this sufficient? Should I use something like this instead?

```
private static void ConfigureAzureIdentityAuthentication(this IServiceCollection services, IConfiguration configuration)
{
    var userAssignedManagedIdentity = configuration.GetValue<string?>("UserAssignedManagedIdentity");

    services.AddSingleton<DefaultAzureCredential>(provider =>
    {
        return new DefaultAzureCredential(new DefaultAzureCredentialOptions
        {
            ManagedIdentityClientId = userAssignedManagedIdentity,
            ExcludeSharedTokenCacheCredential = false
        });
    });

    services.AddAzureClients(clientBuilder =>
    {
        var credential = provider.GetRequiredService<DefaultAzureCredential>();
        clientBuilder.UseCredential(credential);
    });
}

```

This would explain why, after 12 hours of normal operation, all my requests arriving every second try to contact "login.microsoftonline..." at the same time and causing throttling?

**Comment by christothes** [comment]
Can you tell from your telemetry if the excessive calls to login.microsoftonline are coming from a particular client? For example, is it always the SQLClient, or one of the other Azure service clients, or all of them?

If it is working fine for 12 hours it seems that the problem only occurs when the first token refresh is attempted, which usually would occur at the half-life of the token expiration, which is usually 24 hours for managed identity tokens.

It might also be helpful to enable logging for the SDKs to shed some more light on what is happening here:

This can be done as described in [these docs](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/samples/Diagnostics.md#custom-logging-callback).

For example:

```c#
AzureEventSourceListener listener = new AzureEventSourceListener(
    (args, message) => myLogger.Log("[{0:HH:mm:ss:fff}][{1}] {2}", DateTimeOffset.Now, args.Level, message),
    level: EventLevel.Verbose); 
```

Note: you'll want to ensure that the listener does not get disposed for it to persist for the app's lifetime.

**Comment by mhi59** [comment]
It seems to me that this only happens with the SQL client.
In any case, for all the calls I was able to analyze in AppInsights, that was indeed the case.

You’re right, I was wondering about the refresh token since the issue occurs consistently after 12 hours of operation.
To get everything back to normal, I have to restart the application. Then everything works fine again for another 12 hours.

Thank you for your advice regarding the logs; I will implement this on my test environment, where I can reproduce the issue.

For now, in production, I have switched to a login/password connection to avoid the problem.

**Comment by mhi59** [comment]
Hi,

Do you need more informations?

**Comment by christothes** [comment]
Hi @mhi59 - 
What kind of connection string are you using for your SQL client when the problem occurs? If this problem only occurs with SQL Clent, it's possible that there is an issue with the SQLClient is managing its credential lifetime. If so, you may have to open an issue over in [EFCore](https://github.com/dotnet/efcore/issues) or in [SQLClient](https://github.com/dotnet/SqlClient/issues) to investigate what the client is doing when the credential expires.

**Comment by mhi59** [comment]
Hi @christothes ,

The connection string looks like "Server=tcp:sqlsrv-xxx.windows.net,1433;Initial Catalog=sqldb-xxx;Encrypt=True;TrustServerCertificate=False;Connection Timeout=30;Authentication="Active Directory Default";"

Indeed, with login/password auth, this issue disappear. I will take a look on Sql/EF support too.
Thanks for your help.

**Comment by christothes** [comment]
Thanks @mhi59 - this is the connection string I suspected, which is SQLClient's internal implementation for using the `DefaultAzureCredential`. I would recommend asking them how to troubleshoot this, since we have limited control of how to configure logging in this scenario. 

This article explains an alternate approach for using the `DefaultAzureCredential` if you have full access to how the SQLClient is constructed. https://learn.microsoft.com/en-us/dotnet/api/microsoft.data.sqlclient.sqlconnection.accesstokencallback?view=sqlclient-dotnet-standard-5.2

This approach give you full control over the lifetime of the credential, and may help resolve the problem.
